code/__HELPERS/game.dm

```diff
@@ -214,17 +214,29 @@
 /proc/get_hearers_in_view(R, atom/source)
 	// Returns a list of hearers in view(R) from source (ignoring luminosity). Used in saycode.
 	var/turf/T = get_turf(source)
-	var/list/hear = list()
+	. = list()
 
 	if(!T)
-		return hear
-
-	var/list/range = get_hear(R, T)
-	for(var/atom/movable/A in range)
-		hear |= recursive_hear_check(A)
-
-	return hear
+		return
 
+	var/list/processing_list = list()  
+	if (R == 0) // if the range is zero, we know exactly where to look for, we can skip view
+		processing_list += T.contents // We can shave off one iteration by assuming turfs cannot hear
+	else  // A variation of get_hear inlined here to take advantage of the compiler's fastpath for obj/mob in view
+		var/lum = T.luminosity
+		T.luminosity = 6 // This is the maximum luminosity
+		for(var/obj/O in view(R, T))
+			processing_list += O
+		for(var/mob/M in view(R, T))

```

Comment:
Can you replace these with `for(atom/movable/AM in view(R, T))` or is that not optimized by the compiler in the same way?

Reply:
its not, objs and mobs are stored on two different lists on the turf.

Reply:
wait, couldn't this block just be replaced with
```
var/lum = T.luminosity
T.luminosity = 6 // This is the maximum luminosity
processing_list = view(R, T)
T.luminosity = lum
```
?

Reply:
if done this way you can eliminate the call to `list()` up above and just declare `processing_list` without assigning anything, then changing the `+=` in the first block with an `=` as well

Reply:
doing it that way would require us to use |= in the hear_1 check and would kill all performance, because view() returns both mobs, objects and the turfs they are in the contents of, so everything but the turfs would be recursed through twice.

var/mob and var/obj also MAYBE skip lighting objects which derive from AM (though i'm willing to bet "base" AM is actually biased towards one type), further reducing junk.

If you want to reduce this to a single view, var/turf in view is likely the most performant. I will benchmark after i finish my pizza.

Reply:
using
`
		T.luminosity = 6 // This is the maximum luminosity
		for(var/turf/TU in view(R, T))
			processing_list += TU.contents
		T.luminosity = lum
`

| Proc Name                                           | Self CPU | Total CPU | Real Time | Calls |
|-----------------------------------------------------|----------|-----------|-----------|-------|
|/proc/objmobget_hearers_in_view|0.353|0.353|0.353|5517|
|/proc/turfget_hearers_in_view|0.423|0.423|0.421|5710|

If anyone has a good idea on how to do var/thing in view in a way that is more performant than this and doesn't involve the use of |= or type checks, i'm open to suggestions

Reply:
I forgot view() returned turfs.

ya, this is likely to be the fastest. one potential improvement is replace the mob loop. viewers() is like view() but only returns mobs, so you could start with viewers() then do the obj loop.

that is likely to be faster

Reply:
```
		T.luminosity = 6 // This is the maximum luminosity
		for(var/obj/O in view(R, T))
			processing_list += O
		processing_list += viewers(R, T)
		T.luminosity = lum
```

| Proc Name                                           | Self CPU | Total CPU | Real Time | Calls |
|-----------------------------------------------------|----------|-----------|-----------|-------|
|/proc/objmobget_hearers_in_view|0.490|0.490|0.488|6869|
|/proc/viewersget_hearers_in_view|0.558|0.558|0.561|6961|

Reply:
`` 
T.luminosity = 6 // This is the maximum luminosity
		processing_list = viewers(R, T)
		for(var/obj/O in view(R, T))
			processing_list += O
		T.luminosity = lum
`` 

| Proc Name                                           | Self CPU | Total CPU | Real Time | Calls |
|-----------------------------------------------------|----------|-----------|-----------|-------|
|/proc/objmobget_hearers_in_view|0.657|0.657|0.655|8870|
|/proc/viewersget_hearers_in_view|0.608|0.608|0.611|8811|

That's an improvement, checking one more thing and getting final perf after

Reply:
![image](https://user-images.githubusercontent.com/20017308/32751190-9a1b78e8-c8cd-11e7-9919-9861bb45a14e.png)
viewers version has some interesting bugs